replicator.js

```diff
@@ -6,69 +6,64 @@ var properties = [
   {type: 'range', id: 'Spacing (in)', value: 0.5, min: 0, max: 1, step: 0.001},
 ];
 
-// flip a point vertically to the SVG coordinate system where +Y is down
-var convertPointToSvgCoordinates = function(point, height) {
-  return [point[0], height - point[1]];
-};
+var shallowCopy = function(object) {
+  var result = {};
+  for (var nextKey in object) {
+    if (object.hasOwnProperty(nextKey)) {
+      result[nextKey] = object[nextKey];
+    }
+  }
 
-var convertPointArraysToSvgCoordinates = function(pointArrays, height) {
-  return pointArrays.map(function(pointArray) {
-    return pointArray.map(function(point) {
-      return convertPointToSvgCoordinates(point, height);
-    });
-  });
+  return result;
 };
 
-var offsetPoint = function(point, dx, dy) {
-  return [point[0] + dx, point[1] + dy];
-};
+var offsetVolume = function(volume, spacing, x, y) {
+  var newVolume = shallowCopy(volume);
+  var width = volume.shape.width || 0;
+  var height = volume.shape.height || 0;
 
-var offsetPointArrays = function(pointArrays, dx, dy) {
-  return pointArrays.map(function(pointArray) {
-    return pointArray.map(function(point) {
-      return offsetPoint(point, dx, dy);
-    });
-  });
-};
+  newVolume.id = null;
+  newVolume.shape = shallowCopy(volume.shape);
+  newVolume.shape.rotation = newVolume.shape.rotation || 0
+  newVolume.shape.flipping = newVolume.shape.flipping || {}
+  newVolume.shape.center = {
+    x: volume.shape.center.x + (height + spacing) * x,
+    y: volume.shape.center.y + (width + spacing) * y
+  };
 
-var pathSvgDataForPointArray = function(pointArray) {
-  return 'M' + pointArray.map(function(point) { return point[0] + ',' + point[1]; }).join(' ');
-};
+  return newVolume;
+}
 
-var pathSvgForPointArrays = function(pointArrays) {
-  return '<path d="' + pointArrays.map(pathSvgDataForPointArray).join(' ') + '" />';
-};
+var getVolumeById = function(volumes, id) {
+  for (var i = 0; i < volumes.length; i++) {
+    if (volumes[i].id === id) {
+      return volumes[i]
+    }
+  }
+}
 
 // Define a function named `executor` that generates a valid SVG document string
 // and passes it to the provided success callback, or invokes the failure
 // callback if unable to do so
 var executor = function(args, success, failure) {
-  var params = args[0];
+  var params = args.params;
   var columnCount = params['Columns'];
   var rowCount = params['Rows'];
   var spacing = params['Spacing (in)'];
-  var shapeProperties = args[1]; // NOTE: This requires checking "Requires selected shape" on the Edit App page
-  var shapeWidth = shapeProperties.right - shapeProperties.left;
-  var shapeHeight = shapeProperties.top - shapeProperties.bottom;
-  var pointArrays = offsetPointArrays(shapeProperties.pointArrays, -shapeProperties.left, -shapeProperties.bottom);
-  var resultPointArrays = [];
+
+  var selectedVolumeIds = args.selectedVolumeIds || []
+
+  if (selectedVolumeIds.length != 1) {
+    failure('A single volume must be selected!');
```

Comment:
Does this error get presented to the user?

Reply:
![screen shot 2018-08-20 at 3 55 07 pm](https://user-images.githubusercontent.com/596798/44366154-70886400-a491-11e8-84b6-ce96b46d5f8d.png)

Reply:
I think we should use "shape" instead of "volume" then since I don't think Easel uses "volume" in user-facing contexts.